### PR TITLE
Feature/login

### PR DESCRIPTION
**Pull request by tomklecha**
Hi
Managed to properly create project.
Added base classes to the module, next step to work out the inside of classes.
SourceTree is working now

### CODE REVIEW

File: features/login/src/androidTest/java/com/can_apps/login/ExampleInstrumentedTest.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+import org.junit.Assert.*
+
+/**
+ * Instrumented test, which will execute on an Android device.
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+@RunWith(AndroidJUnit4::class)
+class ExampleInstrumentedTest {
```
Comment (Canato) [inline features/login/src/androidTest/java/com/can_apps/login/ExampleInstrumentedTest.kt:17]: delete this

File: features/login/src/main/res/layout/fragment_login.xml [hunk 2]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (Canato) [inline features/login/src/main/res/layout/fragment_login.xml:2]: This is a simple page, we don't need to user Constraint Layout, let's use something with a better performance, like linear layout

File: features/login/src/main/res/layout/fragment_login.xml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:layout_margin="8dp">
+
+    <EditText
+        android:id="@+id/login_edittext"
```
Comment (Canato) [inline features/login/src/main/res/layout/fragment_login.xml:10]: normally we use this notation with underscore to name a xml file, but in kotlin we can call this name directly on our code, so suggestion to use camel case

File: features/login/src/main/res/layout/fragment_login.xml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="100dp"
```
Comment (Canato) [inline features/login/src/main/res/layout/fragment_login.xml:13]: should be divisible by 8, so we can follow Material Design

File: features/login/src/test/java/com/can_apps/login/ExampleUnitTest.kt [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import org.junit.Test
+
+import org.junit.Assert.*
+
+/**
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * See [testing documentation](http://d.android.com/tools/testing).
+ */
+class ExampleUnitTest {
```
Comment (Canato) [inline features/login/src/test/java/com/can_apps/login/ExampleUnitTest.kt:12]: delete this class

File: settings.gradle [hunk 6]
```diff
@@ -2,4 +2,5 @@ include ':app',
         ':common',
 
         // Features Module
-        ':features:questions'
\ No newline at end of file
+        ':features:questions'
+include ':features:login'
```
Comment (Canato) [inline settings.gradle:6]: remove the `include`.

do it:
``` 
':features:questions',
':features:login'
```

**Comment by tomklecha** [comment]
Question,
you asked me to do not implement a lot of code in one pull request or in one commit ?
Could you reccoment me some article for Interactor ? because i am a little bit confused on function of presenter and interactor

File: features/login/src/main/res/layout/fragment_login.xml [hunk 7]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:layout_margin="8dp"
+    android:orientation="vertical">
+
+    <EditText
```
Comment (Canato) [inline features/login/src/main/res/layout/fragment_login.xml:10]: Let's use the material design `TextInputLayout` in the place of Edit Text.

I was not planning to focus on UI now, but let's do this right =D

**Comment by Canato** [comment]
@tomklecha please don't `Resolve Conversation` let me do it when I open =D

> you asked me to do not implement a lot of code in one pull request or in one commit ?

PR, I don't care about commit's hahaha we will squash it. You can do something like create a branch `features/login_data` so you can add all about the data layer. And then we open a PR from `..login_data` to `..login` branch, before merge into master

> Could you recommend me some article for Interactor ? because i am a little bit confused on function of presenter and interactor

This is normal. I did not found a good article about it. Maybe you could right a good one in the end of this project =)

Basically 
- Presenter = what will be showed in the View and take care of the order
- Interactor = Business Logic

This difference will come with time. Don't worry about put something in the wrong file right now. I will help you to put in the best place later.

**Comment by tomklecha** [comment]
one test is not passing through, that was from old implementation, will be fixed today

**Comment by tomklecha** [comment]
@Canato MasterCI passed now, Only GitGuardian on same policy is not going through

File: features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    companion object {
+        fun newInstance(): Fragment = LoginFragment()
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+//        TODO presenterLocator
+//        presenter.bind(this)
+//        presenter.onViewCreated()
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt:25]: ? We should not merge a TODO without a good reason.

File: features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        TODO("Not yet implemented")
+    }
+
+    override fun showError(error: String) {
+        TODO("Not yet implemented")
+    }
+
+    override fun showSuccess() {
+        TODO("Not yet implemented")
+    }
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt:51]: Many todo, we cannot merge it to master. Or totally remove or finish it before we merge to master

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+    private lateinit var view: LoginContract.View
+
+
+    override fun bind(view: LoginContract.View) {
+        this.view = view
+    }
+
+    override fun unbind() {
+    }
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:17]: ON bind you give view a value. unbind you should give view no value
`view = null`

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    override fun onLoginClicked(password: String, loginName: String) {
+        val passwordDomain = LoginPasswordDomain(password)
+        val loginNameDomain = LoginNameDomain(loginName)
+        val isPasswordValid = interactor.passwordValidation(passwordDomain)
+        val isLoginValid = interactor.loginNameValidation(loginNameDomain)
+
+        if (isLoginValid && isPasswordValid) {
+            when (interactor.loginUser(loginNameDomain, passwordDomain)) {
+                LoginDomain.Success -> view.showSuccess()
+                LoginDomain.Fail(LoginErrorDomain("flipflops")) -> view.showError("flipflops")
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:36]: "fliflops" is our error message?... 

On code side: 
```
when (interactor.loginUser(loginNameDomain, passwordDomain)) {
                LoginDomain.Success -> view.showSuccess()
                is LoginDomain.Fail -> view.showError(it.error)
```

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+            }
+        } else
+            view.showError("Sorry, something is invalid")
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:39]: Now will bring something trick for you. Let's use some string from our value folder. 

After you find a way to get the string here (R.string.something_wrong), you need to find a way to update the test for it

File: features/login/src/main/java/com/can_apps/login/core/LoginInteractor.kt [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+        return (patternLowerCase.containsMatchIn(loginName.value) &&
+                patternUpperCase.containsMatchIn(loginName.value) &&
+                pattern.matches(loginName.value))
+    }
+
+    override fun loginUser(loginName: LoginNameDomain, password: LoginPasswordDomain): LoginDomain {
+
+        return repository.loginUser(loginName, password)
+    }
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/core/LoginInteractor.kt:36]: Not need, but a more Kotlin way would be
`override fun loginUser(loginName: LoginNameDomain, password: LoginPasswordDomain): LoginDomain = repository.loginUser(loginName, password)`

File: features/login/src/main/res/layout/fragment_login.xml [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+        android:layout_height="wrap_content"
+        android:layout_gravity="center|center_horizontal"
+        android:layout_marginTop="120dp"
+        android:ems="10"
+        android:gravity="center|center_horizontal"
+        android:hint="@string/login_edittext_hint"
+        android:importantForAutofill="no"
+        android:inputType="textPersonName" />
+
+    <EditText
```
Comment (Canato) [inline features/login/src/main/res/layout/fragment_login.xml:22]: Why one is a `TextInputLayout` and the other is a `EditText`?

File: features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt (hunk 8 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt:25]: resolve #11

File: features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt (hunk 9 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt:51]: resolve #11

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 11 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:36]: thank you for that

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 12 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:39]: will do

File: features/login/src/main/java/com/can_apps/login/core/LoginInteractor.kt (hunk 13 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/core/LoginInteractor.kt:36]: i am happy about any changes to more kotlin style (always) - thanks

File: features/login/src/main/res/layout/fragment_login.xml (hunk 14 above)
Comment (tomklecha) [reply features/login/src/main/res/layout/fragment_login.xml:22]: i think before i done them as two edit text, so you asked me to change. but i can fix it

File: features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt (hunk 9 above)
Comment (Canato) [reply features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt:51]: Wrong notation. This code don't solve 11. it will be solved in 11. 

`WIP in 11` or `Will be done in 11` something like this =p

File: features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt (hunk 8 above)
Comment (Canato) [reply features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt:25]: Anyway, we cannot merge it yet. Let's wait until 11 is merged in this branch.

Please put this Branch as a Draft until is ready 😬

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 10 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:17]: Missed that

File: features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt (hunk 9 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt:51]: Will be done in #11

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 11 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:36]: doesnt work - inresolved reference to "it"

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 12 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:39]: required context in LoginPresenter to extract String resource

File: features/login/src/main/java/com/can_apps/login/core/LoginInteractor.kt (hunk 13 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/core/LoginInteractor.kt:36]: done

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 10 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:17]: can't do it - cannot get null reference to non null value - in Question module is done as a job.cancel() -> i am not familliar with job structure

**Comment by Canato** [comment]
Should not be close. Should be Draft, Otherwise how can we keep following what I wrote here?

**Comment by tomklecha** [comment]
could you please explain me on Monday how to put it to draft ?

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 10 above)
Comment (Canato) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:17]: make view nullable

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 12 above)
Comment (Canato) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:40]: Tip: this should be common like CoroutineDispatcherFactory. So every module can add it

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 10 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:17]: hmm, i cannot find any solution for that now. Because lateinit values cannot be assigned as a null, because first must be initialized.
so it might be i will need to implement coroutines first to have job delegate as you have in questions to cancel the job on view.

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 12 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:40]: i just discovered that we have commom module... I will have better look on all of it tomorrow

**Comment by tomklecha** [comment]
@Canato after giving valid password and login, this is the exception going from FirebaseAuth:
The given sign-in provider is disabled for this Firebase project. Enable it in the Firebase console, under the sign-in method tab of the Auth section.

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 12 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:59]: sorry, with this string i still struggle, i dont know which coroutineContext to use to have access to context.getString(...)

**Comment by tomklecha** [comment]
i tried to add context to getPresenter - 
fun getPresenter(): LoginContract.Presenter = LoginPresenter(getInteractor(), getCoroutineDispatcher(), context)

and then in presenter i am able to use this context. Is that a proper way of using context ?
Isnt that too much for Presenter to know ?
My concenre is about presenter responsibility

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 12 above)
Comment (Canato) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt]: I didn't said to use `CoroutineDispatcherFactory` . I said "Should be common like <this other class that is common> 

You are using something without know what it is.

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt [hunk 15]
```diff
@@ -25,13 +39,18 @@ internal class LoginPresenter(
...
 
         val passwordDomain = LoginPasswordDomain(password)
         val loginNameDomain = LoginNameDomain(loginName)
         val isPasswordValid = interactor.passwordValidation(passwordDomain)
         val isLoginValid = interactor.loginNameValidation(loginNameDomain)
 
         if (isLoginValid && isPasswordValid) {
+
             when (val result = interactor.loginUser(loginNameDomain, passwordDomain)) {
                 LoginDomain.Success -> view.showSuccess()
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:58]: this should be in IO thread?

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt [hunk 16]
```diff
@@ -25,13 +39,18 @@ internal class LoginPresenter(
...
                 is LoginDomain.Fail -> view.showError(result.error.value)
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:59]: this should be in IO thread?

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt [hunk 17]
```diff
@@ -40,4 +59,5 @@ internal class LoginPresenter(
             view.showError("Sorry, something is invalid")
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt]: this should be in IO thread?

File: features/login/build.gradle [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+    //Test
+    testImplementation 'junit:junit:4.12'
+    testImplementation "io.mockk:mockk:1.9.3"
+    androidTestImplementation 'androidx.test.ext:junit:1.1.1'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'
+
+    // Firebase Auth
+    implementation 'com.google.firebase:firebase-auth:19.3.0'
+
+    implementation project(':common')
```
Comment (Canato) [inline features/login/build.gradle:59]: Why we need this?

File: features/login/build.gradle [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+        versionName "1.0"
+
+        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
+        consumerProguardFiles 'consumer-rules.pro'
+    }
+
+    buildTypes {
+        release {
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
```
Comment (Canato) [inline features/login/build.gradle:22]: Do you know what proguard is for?

File: features/login/build.gradle [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+        versionCode 1
+        versionName "1.0"
+
+        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
+        consumerProguardFiles 'consumer-rules.pro'
+    }
+
+    buildTypes {
+        release {
+            minifyEnabled false
```
Comment (Canato) [inline features/login/build.gradle:21]: do you know why this is for?

File: features/login/build.gradle [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
+    implementation "org.jetbrains.kotlinx:kotlinx-coroutines-core:1.3.0"
+    implementation "org.jetbrains.kotlinx:kotlinx-coroutines-android:1.3.0"
+
+    //Test
+    testImplementation 'junit:junit:4.12'
+    testImplementation "io.mockk:mockk:1.9.3"
+    androidTestImplementation 'androidx.test.ext:junit:1.1.1'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'
```
Comment (Canato) [inline features/login/build.gradle:54]: where are we using it?

File: features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+
+    override fun close() {
+        activity?.finish()
+    }
+
+    override fun showLoading() {
+    }
+
+    override fun hideLoading() {
+    }
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt:56]: We will ship this without implementation ?

File: features/login/src/main/java/com/can_apps/login/app/LoginServiceLocator.kt [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+
+    fun getPresenter(): LoginContract.Presenter = LoginPresenter(getInteractor(), getCoroutineDispatcher(), context)
+
+    private fun getInteractor(): LoginContract.Interactor = LoginInteractor(getRepository())
+
+    private fun getRepository(): LoginContract.Repository = LoginRepository(getApi())
+
+    private fun getApi(): LoginFirebaseApi = LoginFirebaseApiDefault(FirebaseAuth.getInstance())
+
+    private fun getCoroutineDispatcher(): CoroutineDispatcherFactory = CoroutineDispatcherFactoryDefault()
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/app/LoginServiceLocator.kt:24]: We had not talk about threads and coroutines yet, please wait to start to use it

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+
+    override fun onBackPressed() {
+        view.close()
+    }
+
+    override fun onLoginClicked(password: String, loginName: String) {
+        retrieveData(password, loginName)
+    }
+
+    private fun CoroutineScope.retrieveData(password: String, loginName: String) = launch(dispatcher.IO) {
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:48]: `launch(dispatcher.IO)` what this is doing?

File: features/login/src/main/res/layout/fragment_login.xml [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+
+    <com.google.android.material.textfield.TextInputLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="120dp"
+        android:gravity="center_horizontal">
+
+        <com.google.android.material.textfield.TextInputEditText
+            android:id="@+id/loginNameEditText"
+            android:layout_width="200dp"
```
Comment (Canato) [inline features/login/src/main/res/layout/fragment_login.xml:17]: What will happen with this in different screens?

File: features/login/src/main/res/layout/fragment_login.xml [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+    </com.google.android.material.textfield.TextInputLayout>
+
+    <com.google.android.material.textfield.TextInputLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:gravity="center_horizontal">
+
+        <com.google.android.material.textfield.TextInputEditText
+            android:id="@+id/loginPasswordEditText"
+            android:layout_width="200dp"
```
Comment (Canato) [inline features/login/src/main/res/layout/fragment_login.xml:31]: What will happen with this in different screens?

File: features/login/src/main/res/layout/fragment_login.xml [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+    </com.google.android.material.textfield.TextInputLayout>
+
+    <Button
+        android:id="@+id/loginSubmitButton"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center|center_horizontal"
+        android:layout_marginTop="16dp"
+        android:gravity="center|center_horizontal"
+        android:text="@string/login_button_text" />
```
Comment (Canato) [inline features/login/src/main/res/layout/fragment_login.xml:45]: What is the difference between `layout_gravity` and `gravity`?

Interview question.

File: features/login/src/test/java/com/can_apps/login/bresenter/LoginPresenterTest.kt [hunk 28]
```diff
@@ -0,0 +1,188 @@
...
+    @InjectMockKs
+    private lateinit var presenter: LoginPresenter
+
+    @Before
+    fun setup() {
+        MockKAnnotations.init(this, relaxed = true)
+
+        val unconfinedFactory = CoroutineDispatcherFactoryUnconfined()
+        every { dispatcherFactory.IO } returns unconfinedFactory.IO
+        every { dispatcherFactory.UI } returns unconfinedFactory.UI
```
Comment (Canato) [inline features/login/src/test/java/com/can_apps/login/bresenter/LoginPresenterTest.kt:39]: we should not be using coroutines yet

File: features/login/src/test/java/com/can_apps/login/bresenter/LoginPresenterTest.kt [hunk 29]
```diff
@@ -0,0 +1,188 @@
...
+    fun `GIVEN valid password and loginName, WHEN onLoginClicked, THEN show success`() {
+        //GIVEN
+        val password = "pass"
+        val loginName = "loginName"
+        val passwordDomain = LoginPasswordDomain(password)
+        val loginDomain = LoginNameDomain(loginName)
+
+        every { interactor.loginNameValidation(loginDomain) } returns true
+        every { interactor.passwordValidation(passwordDomain) } returns true
+        every { interactor.loginUser(loginDomain, passwordDomain) } returns LoginDomain.Success
```
Comment (Canato) [inline features/login/src/test/java/com/can_apps/login/bresenter/LoginPresenterTest.kt:79]: Nothing wrong. Just wanna bring your attention for one thing.

You did change the loginName validation rules (now it need to have @.com.. ) 

And this test keep passing, and this is RIGHT

This is exactly what we want on unit tests. Since we mock the behaviour doesn't matter if you make a change in the interactor class this will not affect the presenter class or test. 

*If you don't change the contract (Core Domain)

File: features/login/src/test/java/com/can_apps/login/core/LoginInteractorTest.kt [hunk 30]
```diff
@@ -0,0 +1,229 @@
...
+    @MockK
+    private lateinit var repository: LoginContract.Repository
+
+    @InjectMockKs
+    private lateinit var interactor: LoginInteractor
+
+
+    @Before
+    fun setup() {
+        MockKAnnotations.init(this, relaxed = true)
```
Comment (Canato) [inline features/login/src/test/java/com/can_apps/login/core/LoginInteractorTest.kt:23]: what is broking without relaxed here?

**Comment by Canato** [comment]
Almost forgot. @tomklecha please don't forget to update the CHANGELOG.md file

File: features/login/build.gradle (hunk 18 above)
Comment (tomklecha) [reply features/login/build.gradle:59]: we need module :common to apply coroutinedispatcherfactory interface

File: features/login/build.gradle (hunk 21 above)
Comment (tomklecha) [reply features/login/build.gradle:54]: using unit testing guarantee us TDD - test driven development, to create stable code, which can be persitance on changes

File: features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt (hunk 22 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/app/LoginFragment.kt:56]: no , i must implement this functions - or delete them from Presenter class since not used

File: features/login/src/main/res/layout/fragment_login.xml (hunk 27 above)
Comment (tomklecha) [reply features/login/src/main/res/layout/fragment_login.xml:45]: layout gravity puts whole layout in the desired position - center, gravity puts elements inside the layout in desired position

File: features/login/src/test/java/com/can_apps/login/bresenter/LoginPresenterTest.kt (hunk 29 above)
Comment (tomklecha) [reply features/login/src/test/java/com/can_apps/login/bresenter/LoginPresenterTest.kt:79]: thanks!

File: features/login/src/test/java/com/can_apps/login/core/LoginInteractorTest.kt (hunk 30 above)
Comment (tomklecha) [reply features/login/src/test/java/com/can_apps/login/core/LoginInteractorTest.kt:23]: the further implementation in verify blocks, so tests are ignoring what actually will further create this function, as long as the funciton will be called - not diffrent one

File: features/login/src/main/res/layout/fragment_login.xml (hunk 26 above)
Comment (tomklecha) [reply features/login/src/main/res/layout/fragment_login.xml:31]: that will be scaled proportionally to screen used. thats why we use dp - densiy independent pixel

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 17 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt]: no that must be on Dispatcher.Main thread- to return solution to main view, not the IO, my bad

File: features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt (hunk 15 above)
Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:58]: as below

Comment (tomklecha) [reply features/login/src/main/java/com/can_apps/login/bresenter/LoginPresenter.kt:58]: all view returns must be on main thread, not IO

**Comment by tomklecha** [comment]
@Canato unfortunately for now i must leave coroutines, 
my only solution to implement view as null would be changing the getPresenter implementation to sent current View in constructor. But that would exclude implementing presenter in onAttach, disable binding functionality and probably code would be more vunerable. Also we could not use then lateinit var

File: features/login/build.gradle (hunk 19 above)
Comment (tomklecha) [reply features/login/build.gradle:22]: in basic idea from google documentation proguard is used to shrink the code, but i dont know yet full idea of it working

File: features/login/build.gradle (hunk 21 above)
Comment (Canato) [reply features/login/build.gradle:54]: where are you using espresso test?

File: features/login/src/test/java/com/can_apps/login/core/LoginInteractorTest.kt (hunk 30 above)
Comment (Canato) [reply features/login/src/test/java/com/can_apps/login/core/LoginInteractorTest.kt:23]: I went into the code and remove `relaxed = true` the tests still working.

> what is broking without relaxed here?

Answer: Nothing.

File: settings.gradle [hunk 31]
```diff
@@ -2,5 +2,5 @@ include ':app',
         ':common',
 
         // Features Module
-        ':features',
-        ':features:questions'
\ No newline at end of file
+        ':features:login',
+        ':features:questions'
```
Comment (Canato) [inline settings.gradle]: You remove `:features`

Comment (tomklecha) [reply settings.gradle]: yes, my bad, i thought i have done mistake to put features without login

File: features/login/src/test/java/com/can_apps/login/core/LoginInteractorTest.kt (hunk 30 above)
Comment (tomklecha) [reply features/login/src/test/java/com/can_apps/login/core/LoginInteractorTest.kt]: i know now what was broking before - is it because i added implementation of the view.showError -> and there is a function which do it now ?
before i didnt implement the functionality. just a question for that

**Comment by tomklecha** [comment]
can you explain me your changes/fixes ? now somehow the guardian check is working, so i dont know what happened

File: features/login/src/main/java/com/can_apps/login/core/LoginContract.kt [hunk 32]
```diff
@@ -0,0 +1,49 @@
...
+
+    interface Interactor {
+
+        suspend fun passwordValidation(password: LoginPasswordDomain): Boolean
+
+        suspend fun loginNameValidation(loginName: LoginNameDomain): Boolean
+
+        suspend fun loginUser(loginName: LoginNameDomain, password: LoginPasswordDomain): LoginDomain
+
+        fun logoutUser()
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/core/LoginContract.kt:39]: suspend ;)

File: features/login/src/main/java/com/can_apps/login/data/LoginRepository.kt [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+
+    override suspend fun loginUser(
+        loginName: LoginNameDomain,
+        password: LoginPasswordDomain
+    ): LoginDomain {
+        try {
+            val result = api.loginExistingUser(loginName.value, password.value)
+            if (!result) {
+                val result2 = api.createNewUser(loginName.value, password.value)
+                if (!result2) {
```
Comment (Canato) [inline features/login/src/main/java/com/can_apps/login/data/LoginRepository.kt:15]: remove result and result2
